sharding/sharding.go

```diff
@@ -0,0 +1,311 @@
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/allegro/akubra/config"
+	"github.com/allegro/akubra/httphandler"
+	"github.com/allegro/akubra/transport"
+	"github.com/golang/groupcache/consistenthash"
+)
+
+type cluster struct {
+	http.RoundTripper
+	weight   uint
+	backends []config.YAMLURL
+	name     string
+}
+
+type shardsRing struct {
+	ring                    *consistenthash.Map
+	shardClusterMap         map[string]cluster
+	allClustersRoundTripper http.RoundTripper
+	clusterRegressionMap    map[string]cluster
+	inconsistencyLog        *log.Logger
+}
+
+func (sr shardsRing) isBucketPath(path string) bool {
+	trimmedPath := strings.Trim(path, "/")
+	return len(strings.Split(trimmedPath, "/")) == 1
+}
+
+func (sr shardsRing) Pick(key string) (cluster, error) {
+	var shardName string
+
+	shardName = sr.ring.Get(key)
+	shardCluster, ok := sr.shardClusterMap[shardName]
+	if !ok {
+		return cluster{}, fmt.Errorf("no cluster for shard %s, cannot handle key %s", shardName, key)
+	}
+
+	return shardCluster, nil
+}
+
+type reqBody struct {
+	r *bytes.Reader
+}
+
+func (rb *reqBody) rewind() error {
+	_, err := rb.r.Seek(0, io.SeekStart)
+	return err
+}
+
+func (rb *reqBody) Read(b []byte) (int, error) {
+	return rb.r.Read(b)
+}
+
+func (rb *reqBody) Close() error {
+	return nil
+}
+
+func copyRequest(origReq *http.Request) (*http.Request, error) {
+	newReq := new(http.Request)
+	*newReq = *origReq
+	newReq.URL = &url.URL{}
+	*newReq.URL = *origReq.URL
+	newReq.Header = http.Header{}
+	for k, v := range origReq.Header {
+		for _, vv := range v {
+			newReq.Header.Add(k, vv)
+		}
+	}
+	if origReq.Body != nil {
+		buf := new(bytes.Buffer)
+		_, err := io.Copy(buf, origReq.Body)
+		if err != nil {
+			return nil, err
+		}
+		newReq.Body = &reqBody{bytes.NewReader(buf.Bytes())}
+	}
+	return newReq, nil
+}
+
+func (sr shardsRing) send(roundTripper http.RoundTripper, req *http.Request) (*http.Response, error) {
+	// Rewind request body
+	bodySeeker, ok := req.Body.(*reqBody)
+	if ok {
+		err := bodySeeker.rewind()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return roundTripper.RoundTrip(req)
+}
+
+func (sr shardsRing) regressionCall(cl cluster, req *http.Request) (string, *http.Response, error) {
+	resp, err := sr.send(cl, req)
+	// Do regression call if response status is > 400
+	if err != nil || resp.StatusCode > 400 {
+		rcl, ok := sr.clusterRegressionMap[cl.name]
+		if ok {
+			return sr.regressionCall(rcl, req)
+		}
+	}
+	return cl.name, resp, err
+}
+
+func (sr shardsRing) RoundTrip(req *http.Request) (*http.Response, error) {
+	reqCopy, err := copyRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if reqCopy.Method == http.MethodDelete || sr.isBucketPath(reqCopy.URL.Path) {
+		return sr.allClustersRoundTripper.RoundTrip(reqCopy)
+	}
+
+	cl, err := sr.Pick(reqCopy.URL.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	clusterName, resp, err := sr.regressionCall(cl, reqCopy)
+	if clusterName != cl.name {
```

Comment:
Extract body of this `if` to function e.g., `logInconsistency(reqCopy.URL.Path, cl.name, clusterName)`